### PR TITLE
[Feature:InstructorUI] Teaching Staff Locked Gradeables Access

### LINKED ISSUE #8355: "Locked" gradeables are not visible to Teaching Staff
**What problem are you trying to solve with Submitty**
The feature to 'lock' a gradeable until another gradeable is completed satisfactorily can be a burden to teaching staff.   Teaching staff cannot view or submit to the later gradeable -- which can block their ability to help a student who has questions about the later gradeable.

**Describe the way you'd like to solve this problem**
Teaching staff (limited, full access, and instructor) should not be blocked from viewing these gradeables.  I would still like it to be clearly labeled that the gradeable does have the dependency on another gradeable, but maybe after clicking to confirm that they want to proceed, the teaching staff can access the gradeable.

### PR DESCRIPTION
**Pull request by ch3nnyj**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?

Currently locked gradeables were previously inaccessible for teaching staff and instructors. 
This would prevent students from getting help with certain gradeables if teaching staff hasn't completed the prerequisite.
When teaching staff attempts to access locked graded the below alert pops up.
Screenshot 2023-02-28 at 2 25 36 PM
Closes #8355 
### What is the new behavior?

Now locked gradeables can be bypassed with a confirmation screen allowing teaching staff to access the submission page.
After clicking ok teaching staff will be redirected to the submission page.
Screenshot 2023-02-28 at 3 23 19 PM

### Other information?
To Test:
1.) Create a gradeable that requires another gradable as a prerequisite.
2.) Attempt to access the as a student and verify that there is no access to that gradeable.
3.) Verify all teaching staff (Mentors/TAs/Instructors) can access the gradeable after click ok on the conformation prompt.

### CODE REVIEW

**Review by jason490 — CHANGES_REQUESTED** [review]
Change '==' to '===' for strict equality

**Comment by williamjallen** [comment]
> Change '==' to '===' for strict equality

@jason490 A quick comment about reviewing: you should mark the specific lines (or at least one example of a problematic line) when you suggest changes like this.  It makes it easier to see what you're describing without clicking into the files pane.  Out of the 47 lines added, only a handful have the `==` operator, so it's not immediately obvious which portion of the changes you're referencing.

To comment on specific lines, hover over the changed line in the files panel and click the little blue plus sign.  Doing so will allow you to add comments on specific lines in addition to the overall comment.
image

File: site/app/templates/officeHoursQueue/CurrentQueue.twig [hunk 1]
```diff
@@ -24,7 +24,7 @@
 
           {% if entry['current_state'] == 'waiting' %}
             <td>
-              <form method="post" action="{{base_url}}/{{entry['queue_code'] | url_encode }}/startHelp" {% if entry['paused'] %}onsubmit="return confirm('Are you sure you want to start helping this student? This student is currently paused.');"{% endif %}>
+             <form method="post" action="{{base_url}}/{{entry['queue_code'] | url_encode }}/startHelp" {% if entry['paused'] %}onsubmit="return confirm('Are you sure you want to start helping this student? This student is currently paused.');"{% endif %}>
```
Comment (williamjallen) [inline site/app/templates/officeHoursQueue/CurrentQueue.twig:27]: This change seems unrelated and should be reverted.

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/app/templates/officeHoursQueue/CurrentQueue.twig (hunk 1 above)
Comment (ch3nnyj) [reply site/app/templates/officeHoursQueue/CurrentQueue.twig:27]: Ah yes, I changed this when messing around and trying to do the 3 variable solution and forgot to change it back when I removed it.

File: site/app/templates/Navigation.twig [hunk 2]
```diff
@@ -92,15 +92,31 @@
...
     {% else %}
         <a class="{{ button.getClass() }}{{ (button.getTitle() == 'Edit' or button.getTitle() == 'Delete') and has_build_error ? 'edit-build-error' : '' }} {% if button.hasOnclick() %} key_to_click {% endif %}" tabindex="0"
             {% if button.isDisabled() %}
-                disabled
-                onclick="alert('Please complete {{ button.getPrerequisite()}} first' )"
-                title='Please complete {{ button.getPrerequisite()}} first'
+                {% if button.getBypass() %}
+                    disabled
+                    {% set str1 = 'This gradeable requires ' %}
+                    {% set str2 = ' [PLACEHOLDER TEXT HERE]'%}
```
[the remark this answers is not in the document]
Comment (ch3nnyj) [reply site/app/templates/Navigation.twig:112]: Waiting on Barb to get back and let me know what to put in.

File: site/app/models/gradeable/Gradeable.php [hunk 3]
```diff
@@ -2357,26 +2357,33 @@ public function setAllowedMinutesOverrides(array $overrides): void {
      * Determines if gradeable is locked for user
      *
      * @param string $user_id
-     * @return bool
+     * @return int
+     *
+     * 0 = False, 1 = True, 2 = True but allows for change
```
[the remark this answers is not in the document]
Comment (ch3nnyj) [reply site/app/models/gradeable/Gradeable.php:2362]: I will definitely look into it and find a better way, after looking at it I think its possible to make it remain a boolean

**Comment by AdamTuhacek** [comment]
I did some testing, both on student and instructor view, and getting below the threshold and then above. Everything seems to be working as it should. Maybe I should take another look when the changes are finalized.

**Review by jason490 — CHANGES_REQUESTED** [review]
image
It seems like there is this pop up box that shows when I tried submitting something into the locked gradable. I tried using the different type of files and it seems like it still shows up. It seems like something is triggering this pop up to show

**Review by jason490 — COMMENTED** [review]
I also have another issue where Im not able to re-lock the gradeable after I unlock it. Though this is not a problem in this PR because this seems to have the same issue on the main branch

File: site/app/views/NavigationView.php [hunk 4]
```diff
@@ -591,8 +591,12 @@ public static function getSubmitButton(Core $core, Gradeable $gradeable, $graded
                 }
             }
         }
+        $bypass = false;
         $prerequisite = '';
-        if ($gradeable->isLocked($core->getUser()->getId())) {
+        if ($gradeable->isLocked($core->getUser()->getId()) == 1 || $gradeable->isLocked($core->getUser()->getId()) == 2) {
+            if ($gradeable->isLocked($core->getUser()->getId()) == 2) {
```
[the remark this answers is not in the document]
Comment (ch3nnyj) [reply site/app/views/NavigationView.php]: I think these changes were outdated. I no longer use 1 and 2, I've switched to booleans. Has the branch not be merged with my commits?

**Comment by ch3nnyj** [comment]
> the

I will look into it.

**Comment by ch3nnyj** [comment]
@williamjallen Can you review this again

File: site/app/controllers/student/SubmissionController.php [hunk 5]
```diff
@@ -114,7 +114,7 @@ public function showHomeworkPage($gradeable_id, $gradeable_version = null) {
             return $verify_permissions;
         }
 
-        if ($gradeable->isLocked($this->core->getUser()->getId())) {
+        if ($gradeable->isLocked($this->core->getUser()->getId()) && ($this->core->getUser()->accessGrading() === false)) {
```
Comment (williamjallen) [inline site/app/controllers/student/SubmissionController.php:117]: These parentheses are unnecessary
```suggestion
        if ($gradeable->isLocked($this->core->getUser()->getId()) && $this->core->getUser()->accessGrading() === false) {
```

File: site/app/controllers/student/SubmissionController.php [hunk 6]
```diff
@@ -1069,7 +1069,7 @@ public function ajaxUploadSubmission($gradeable_id, $merge = null, $clobber = nu
             return $this->uploadResult($msg, false);
         }
 
-        if ($gradeable->isLocked($user_id)) {
+        if ($gradeable->isLocked($user_id) && (!$this->core->getUser()->accessGrading())) {
```
Comment (williamjallen) [inline site/app/controllers/student/SubmissionController.php:1072]: Same here

```suggestion
        if ($gradeable->isLocked($user_id) && !$this->core->getUser()->accessGrading()) {
```

**Review by williamjallen — CHANGES_REQUESTED** [review]
I think this PR is over-complicating the solution.  Instead of having a bypass attribute, I think you an just make a simple check in the relevant Twig file to check the user's access level.

File: site/app/models/Button.php [hunk 7]
```diff
@@ -69,7 +69,6 @@ class Button extends AbstractModel {
     protected $icon;
      /** @prop @var string|null $prefix */
     protected $prefix;
-
```
Comment (williamjallen) [inline site/app/models/Button.php:72]: This change seems to be unrelated to the rest of this PR and should be reverted

File: site/app/models/gradeable/Gradeable.php [hunk 8]
```diff
@@ -2358,6 +2358,7 @@ public function setAllowedMinutesOverrides(array $overrides): void {
      *
      * @param string $user_id
      * @return bool
+     *
```
Comment (williamjallen) [inline site/app/models/gradeable/Gradeable.php:2361]: Same here.  This change is unrelated to the rest of this PR.  Whitespace changes like this can unintentionally cause merge conflicts in other PRs.

File: site/app/views/NavigationView.php [hunk 9]
```diff
@@ -607,7 +607,7 @@ public static function getSubmitButton(Core $core, Gradeable $gradeable, $graded
             "disabled" => $disabled,
             "class" => "btn {$class} btn-nav btn-nav-submit",
             "name" => "submit-btn",
-            "prerequisite" => $prerequisite
+            "prerequisite" => $prerequisite,
```
Comment (williamjallen) [inline site/app/views/NavigationView.php:610]: Same here.

File: site/public/js/server.js [hunk 10]
```diff
@@ -1096,6 +1096,12 @@ function enableTabsInTextArea(jQuerySelector) {
     });
 }
 
+function confirmBypass(str, redirect){
+    confirmed = confirm(str);
+    if (confirmed)
+        location.href = redirect;
```
Comment (williamjallen) [inline site/public/js/server.js:1102]: Submitty's stye guide requires parentheses on all if-statements.  It's also generally better practice, since future developers will have to edit fewer lines if they want to add more lines to the body of the statement.

```suggestion
    if (confirmed) {
        location.href = redirect;
    }
```

File: site/public/js/server.js [hunk 11]
```diff
@@ -1096,6 +1096,12 @@ function enableTabsInTextArea(jQuerySelector) {
     });
 }
 
+function confirmBypass(str, redirect){
```
Comment (williamjallen) [inline site/public/js/server.js:1099]: This file is not linted by `eslint`, but the linter would complain about this.

```suggestion
function confirmBypass(str, redirect) {
```

File: site/public/js/server.js [hunk 12]
```diff
@@ -1096,6 +1096,12 @@ function enableTabsInTextArea(jQuerySelector) {
...
+    confirmed = confirm(str);
+    if (confirmed)
```
Comment (williamjallen) [inline site/public/js/server.js:1101]: `confirmed` needs to be properly declared with `let` or `const`.  Alternatively, you could just move `confirm(str)` to the body of the if-statement.

**Review by williamjallen — DISMISSED** [review]

**Review by williamjallen — COMMENTED** [review]
Requested changes have been made.  I'll leave the functionality review to someone else for final approval.

**Comment by ch3nnyj** [comment]
@williamjallen Can you just approve the changes when you get a chance @jason490 verified the functionality

**Review by jason490 — APPROVED** [review]
Works, but make sure the prompt text changes

File: site/app/templates/Navigation.twig [hunk 13]
```diff
@@ -98,9 +98,17 @@
     {% else %}
         <a class="{{ button.getClass() }}{{ (button.getTitle() == 'Edit' or button.getTitle() == 'Delete') and has_build_error ? 'edit-build-error' : '' }} {% if button.hasOnclick() %} key_to_click {% endif %}" tabindex="0"
             {% if button.isDisabled() %}
-                disabled
-                onclick="alert('Please complete {{ button.getPrerequisite()}} first' )"
-                title='Please complete {{ button.getPrerequisite()}} first'
+                {% if core.getUser().accessFullGrading() %}
+                    disabled
+                    {% set str1 = 'This gradeable requires ' %}
+                    {% set str2 = ' Waiting on Barb to give me a prompt!!!'%}
```
[the remark this answers is not in the document]
Comment (ch3nnyj) [reply site/app/templates/Navigation.twig:104]: @bmcutler Can you give me a prompt to fill this in?

**Comment by williamjallen** [comment]
Oops, I think this might have been merged too early.  There is still a placeholder string in the code.

**Comment by ch3nnyj** [comment]
> Oops, I think this might have been merged too early. There is still a placeholder string in the code.

I already informed Barb and she will change it, I talked to her during the meeting today. I just have to remind her everyday at 9 am. :)
